chore(ui): migrate Message component to Typescript

guoda-puidokaite opened the pull request:
# Summary

- Migrated the `Message` component to `TypeScript`, enhancing code quality, maintainability, and scalability.
- Utilised static typing to minimize runtime errors and improve the development experience.
- Refactored the component for cleaner, more maintainable code.

# Changes Made

<!-- List the changes that were made in this pull request. -->

**TypeScript Migration**:

- Converted `Message` component files to `TypeScript`.
- Added comprehensive type annotations.

**Bug Fix:**

- Auto-dismiss is now correctly triggered only for values greater than or equal to one.

**Refactoring**:

- Made necessary adjustments to address TypeScript-detected issues.
- Further refactored for cleaner and more understandable code.

**Testing**:

- Added tests for bug fix.
- Fixed tests that weren't checking for the close button in case of duplicates.
- Organised tests by what they check for better readability and maintenance.
- Verified component behaviour using `Storybook`.
- Updated tests to use `vitest`.

**Cleanup:**

- Removed deprecated code. It wasn't needed for any `.js` components.

**Documentation:**

- Updated `Storybook` documentation.

Screenshot: Deprecated `.js` code on the left, migrated `.ts` code on the right (**see notes**)

<img width="1728" alt="Screenshot 2024-10-14 at 08 13 31" src="https://github.com/user-attachments/assets/95e0b0c8-190d-4e08-947d-52d6f67f8f0a">

**Release:**

- Generated a `changeset` file, mentioning the migration and bug fix.

# Related Issues

<!-- List any related issues or tickets, including links to them. -->

- https://github.com/cloudoperators/juno/issues/238

# Testing Instructions

<!-- Describe the steps needed to test this pull request. -->

1. Run `Storybook`
2. Run `Message` tests

# Checklist

<!-- Ensure that your pull request meets the following requirements. -->

- [x] I have performed a self-review of my code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have made corresponding changes to the documentation (if applicable).
- [x] My changes generate no new warnings or errors.

guoda-puidokaite reviewed (commented):
Bug found: A timeout of zero or a negative number initiates a message dismissal. [Mentioned in main issue](https://github.com/cloudoperators/juno/issues/238#issuecomment-2413302523).

andypf commented on packages/ui-components/src/components/Message/Message.types.ts:
```diff
@@ -0,0 +1,27 @@
+  /** Optional timeout in milliseconds after which the message auto-dismisses. Default is 10000 (10 seconds). */
+  autoDismissTimeout?: number
+  /** Optional handler function called once the message is dismissed */
+  onDismiss?: () => void
```
This results in the type being displayed in a user-unfriendly way in Storybook.

![image](https://github.com/user-attachments/assets/0c507f2f-f648-4caf-adc7-c9ff81dc0b10)

I’m not an expert in TypeScript yet, but could `React.MouseEventHandler` be useful in this case?

edda replied:
Check e.g. the `Button` component. It has been moved to TS and also has click event handlers

guoda-puidokaite replied:
Thank You @andypf and @edda.
I think that `React.MouseEventHandler` would be incorrect in this context. 
In the `Button` component, it's tied to the specific mouse event `onClick`, so `React.MouseEventHandler<EventTarget>` is correct.
But here, the `onDismiss` function is a custom handler. It should just be a function that takes no arguments and returns nothing - `(() => void)`.

andypf replied:
Got it, let’s wait until Wowa is here, maybe he has an idea.

andypf reviewed (commented):
Good Job Guoda 👍. However, I found a couple of issues.

andypf commented:
> Bug found: A timeout of zero or a negative number initiates a message dismissal

Did you fix it?

edda commented on packages/ui-components/src/components/Message/Message.component.tsx:
```diff
@@ -0,0 +1,79 @@
+import { Icon } from "../Icon"
+
+import { MessageProps } from "./Message.types"
+import { getMuiIcon } from "./utils/iconUtils"
```
I don't see a big advantage of moving the icon logic into utils if only this one component makes use of it. However with small adjustments it might be reusable for all components that use icons. So instead of checking `MessageType` make a more generic type like `SemanticVariant` or something, then move the iconUtils up a few levels to `src/utils`. 

Though my suggestion would be to not do the generalization now but rather do this in a new PR and then use this new shared util in all components with icons. Though it might be more useful to do this after everything (or at least all components that use icons) have moved to Typescript :)

guoda-puidokaite replied:
I didn't want to touch the other components and just try it out here. 🙈 But I'll revert all externalisation and we can discuss it in the future. Thank You.

edda commented on packages/ui-components/src/components/Message/Message.component.tsx:
```diff
@@ -0,0 +1,79 @@
+
+import { MessageProps } from "./Message.types"
+import { getMuiIcon } from "./utils/iconUtils"
+import { initiateAutoDismiss, clearAutoDismissTimeout } from "./utils/timeoutUtils"
```
I would prefer to keep this inline in the `Message` component right now. Mostly to be in line with other similar components where it is like this, too. It makes it harder to see what the functions do because you have to jump to another file. 

In general I'm not against this type of refactoring but then we need to look at it holistically and do the same thing in all components where it applies to ensure consistency.

guoda-puidokaite replied:
Thank You for the review, Esther. No problem, I can do that. For me, it makes the component code very hard to read. I like to open a component file and see quickly what it does. Other helpers, especially css variables etc. are secondary to that, in my opinion. 

Overall it was a really nice way for me to play around with the code and get familiar with it. 🤪

But the most important thing indeed is consistency. If the team prefers to structure code this way across all components and find it easier to understand the full logic in one place, I agree we should keep it this way. Thank You. 🚀

guoda-puidokaite commented on packages/ui-components/src/components/Message/Message.component.tsx:
```diff
@@ -0,0 +1,79 @@
+import { MessageProps } from "./Message.types"
+import { getMuiIcon } from "./utils/iconUtils"
+import { initiateAutoDismiss, clearAutoDismissTimeout } from "./utils/timeoutUtils"
+import {
```
Will do. I'd at the very least consider placing the main component at the top, to make it easier to quickly understand the core logic of the file without scrolling past utility functions and type definitions. 🙈 And due to beauty of hoisting still be good. But again convention, so we can leave it as is.

guoda-puidokaite commented on packages/ui-components/src/components/Message/Message.types.ts:
```diff
@@ -0,0 +1,27 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+export type MessageType = "info" | "warning" | "danger" | "error" | "success"
```
@edda Should we move types into the main component? Other components do so.

guoda-puidokaite replied:
Done

guoda-puidokaite commented:
Hey, @andypf and @edda. I made the changes and added comments for everything else. Thank You for the review. 🚀

Changes made:
- Bug fixed: Auto dismiss is triggered for values less than one
- Added tests for fix and grouped tests by type
- Updated the `changeset` message to include bug fix
- Moved all externalised `css`, `icon` and `timeout` logic back to the main component

andypf reviewed (dismissed):
LGTM

barsukov commented on packages/ui-components/src/components/Message/Message.test.tsx:
```diff
@@ -0,0 +1,162 @@
+    expect(screen.getByRole("img")).toHaveClass("jn-text-theme-info")
+  })
+
+  test.each`
```
i would like to discuss this point it is very similar to a Badge PR :)

barsukov commented on packages/ui-components/src/components/Message/Message.test.tsx:
```diff
@@ -0,0 +1,162 @@
+  test("renders a dismissible Message when 'dismissible' is true", async () => {
+    render(<Message data-testid="my-message" dismissible={true} />)
+
+    const user = userEvent.setup()
```
Do we need a setup here?

guoda-puidokaite replied:
We probably don't need it for such a simple test, though it's recommended for more reliable and accurate test results. It helps handle asynchronous state updates and side effects. But just fyi: We use it across our code base, I didn't add it here. 🤪

barsukov commented on packages/ui-components/src/components/Message/Message.test.tsx:
```diff
@@ -0,0 +1,162 @@
+    expect(screen.getByTestId("my-message")).toBeInTheDocument()
+
+    // Wait to check that the message did not auto-dismiss
+    await new Promise((resolve) => setTimeout(resolve, 1000))
```
Do we a promise here? is not waitFor or act are made for?

guoda-puidokaite replied:
Hmmm, I don't think so. Added due to this error (`waitFor` was there without `await`): 
<img width="717" alt="Screenshot 2024-10-14 at 14 00 57" src="https://github.com/user-attachments/assets/7b161967-0eda-407b-9839-ac17348c9013">
To my understanding, `waitFor` repeatedly calls what's inside until the condition is met or the timeout is reached. 
But it's still a `Promise`, so we need to stop execution in our code for it to be fulfilled.
But good point, now that I reflect on it... Can't we just do below? Much more readable.
```
await user.click(screen.getByRole("button", { name: /close/i }))
expect(screen.queryByTestId("my-message")).not.toBeInTheDocument()
```
However, the first way is used all over our code base...

barsukov commented on packages/ui-components/src/components/Message/Message.component.tsx:
```diff
@@ -0,0 +1,210 @@
+
+import { MessageProps, VariantType } from "./Message.types"
+
+// CSS UTILS
```
Lets discuss this component on a call.

barsukov reviewed (changes requested):
Good job! I have few comments left, lets jump on a call and discuss :) 🔢

barsukov commented on packages/ui-components/src/components/Message/Message.component.tsx:
```diff
@@ -0,0 +1,259 @@
+  successBg: `jn-bg-theme-message-success`,
+}
+
+export type VariantType = "info" | "warning" | "danger" | "error" | "success"
```
I would like to rename it here same way as we did with badge

guoda-puidokaite replied:
Ah, good catch!

guoda-puidokaite commented on packages/ui-components/src/components/Message/Message.component.tsx:
```diff
@@ -0,0 +1,259 @@
+// Determine if a given string corresponds to a known icon
+const isValidIcon = (icon: string): icon is KnownIcons => {
+  const validIconNames: Set<KnownIcons> = new Set(Object.values(KnownIconsEnum))
+  return validIconNames.has(icon as KnownIcons)
```
Yes, just because it's a `string`. But because we know the result will be valid `KnownIcon`, should be fine to cast.

barsukov reviewed (approved):
Great! 🥇

andypf reviewed (approved):
👍

edda reviewed (approved):
🚀
